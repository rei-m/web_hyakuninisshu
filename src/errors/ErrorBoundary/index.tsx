import * as React from 'react';
import Img from 'gatsby-image';
import MessagePageTemplate from '@src/components/templates/MessagePageTemplate';
import Txt from '@src/components/atoms/Txt';
import { useDogezaImage } from '@src/hooks/staticQueries/useDogezaImage';
import { AppError, UNKNOWN_MESSAGE } from '@src/errors';

export interface State {
  error?: Error;
}

export default class ErrorBoundary extends React.Component<{}, State> {
  constructor(props: {}) {
    super(props);
    this.state = {
      error: undefined,
    };
  }

  public componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    if (process.env.NODE_ENV !== 'production') {
      console.error(errorInfo);
    }
    this.setState({
      error,
    });
  }

  public render() {
    const { error } = this.state;
    if (error) {
      const message = error instanceof AppError ? error.message : UNKNOWN_MESSAGE;
      return (
        <MessagePageTemplate>
          <Txt size={`l`}>{message}</Txt>
          <Img
            fluid={useDogezaImage()}
            style={{
              width: 200,
              marginTop: '32px',
            }}
            alt={`エラーが起きました`}
          />
        </MessagePageTemplate>
      );
    }
    return this.props.children;
  }
}
