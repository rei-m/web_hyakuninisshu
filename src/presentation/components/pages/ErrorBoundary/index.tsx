import React from 'react';
import UnhandledErrorPage from '@src/presentation/components/pages/500';

export type State = {
  error?: Error;
};

export default class ErrorBoundary extends React.Component<{}, State> {
  constructor(props: {}) {
    super(props);
    this.state = {
      error: undefined,
    };
  }

  public componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    if (process.env.NODE_ENV !== 'production') {
      console.error(errorInfo);
    }
    this.setState({
      error,
    });
  }

  public render() {
    const { error } = this.state;
    if (error) {
      return <UnhandledErrorPage error={error} />;
    }
    return this.props.children;
  }
}
