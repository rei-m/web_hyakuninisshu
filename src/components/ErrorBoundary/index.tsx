import * as React from 'react';
import styled from 'styled-components';
import { appTheme, withAppTheme } from '@src/styles';
import { AppError, UNKNOWN_MESSAGE } from '@src/errors';
import Dogeza from '@src/components/Dogeza';

export interface ErrorBoundaryState {
  readonly error?: Error;
}

const Root = withAppTheme(styled.div)`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme }) => theme.colorThin};
  flex-direction: column;
  box-sizing: border-box;
  height: 100vh;
`;

const Message = styled.div`
  font-size: 1.8rem;
`;

export default class ErrorBoundary extends React.Component<
  {},
  ErrorBoundaryState
> {
  constructor(props: {}) {
    super(props);
    this.state = {
      error: undefined
    };
  }

  public componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    if (process.env.NODE_ENV !== 'production') {
      console.error(errorInfo);
    }
    this.setState({
      error
    });
  }

  public render() {
    const { error } = this.state;
    if (error) {
      const message =
        error instanceof AppError ? error.message : UNKNOWN_MESSAGE;
      return (
        <Root>
          <Message>{message}</Message>
          <Dogeza
            alt="エラーが起きました"
            style={{
              marginTop: appTheme.spacing4x
            }}
          />
        </Root>
      );
    }
    return this.props.children;
  }
}
