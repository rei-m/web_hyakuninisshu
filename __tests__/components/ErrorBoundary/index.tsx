import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import ErrorBoundary from '@src/components/ErrorBoundary';

describe('<ErrorBoundary />', () => {
  it('should render children', () => {
    const renderer = ReactTestRenderer.create(
      <ErrorBoundary>test</ErrorBoundary>
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render error', () => {
    const renderer = ReactTestRenderer.create(
      <ErrorBoundary>test</ErrorBoundary>
    );
    renderer.root.instance.setState({ error: new Error('hoge') });
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
