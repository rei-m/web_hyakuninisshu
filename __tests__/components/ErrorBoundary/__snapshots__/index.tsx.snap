// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> should render children 1`] = `"test"`;

exports[`<ErrorBoundary /> should render error 1`] = `
<div
  className="sc-bwzfXH bTlVnA"
>
  <div
    className="sc-htpNat fSNmNw"
  >
    予期せぬエラーが発生しました。時間をおいて再度お試しください。
  </div>
  <img
    alt="エラーが起きました"
    className="sc-bdVaJa dxKQNh"
    src="dogeza_businessman.png"
    style={
      Object {
        "marginTop": "32px",
      }
    }
  />
</div>
`;
